docs/types/dynamic-dispatch.md

```diff
@@ -162,3 +162,137 @@ main = Test.simplify
 When invoking a method on _module object_ its _module static methods_ take
 precedence over _instance methods_ defined on `Any`. Thus a module serves
 primarily as a _container for module (static) methods_.
+
+## Method Resolution
+
+This section describes the _method resolution_ process, which resolves a
+concrete method definition for a concrete call site. For a method call
+expression `Receiver.symbol`, this section focuses only on a single dispatch
+based on `Receiver` argument. For multiple dispatch, see the
+[Multiple Dispatch](#multiple-dispatch) section.
+
+Method resolution algorithm for `Receiver.symbol`:
+
+1. **Determine the kind of `Receiver`:**
+
+   - 1.1. If `Receiver` is a type (type object), go to step 2.
+   - 1.2. If `Receiver` is a value (instance / atom), go to step 3.
+   - 1.3. If `Receiver` is a module, go to step 4.
+   - 1.4. If `Receiver` is a polyglot object, go to step 5.
+   - 1.5. If there is no `Receiver`, go to step 6.
+
+2. **`Receiver` is a type. This call site is a _static method call_:**
```

Comment:
To play devil's advocate: 
- Isn't it all a bit different?
- The 1st step is to find a `Type` for the receiver
   - for `Atom` the type is its defining type
   - for module the type is its associated type
   - for a `Type` it is its _eigen type_
   - etc.?
- then lookup the method in the _pool_ of that type or
   - its `getSupertype` pool and so on
- if a method is found, invoke it with `self` being set to the receiver
- if no such a method is found, consider _static invocation of instance method_!?
   - _and somewhere here my thoughts dry out..._

Reply:
> _and somewhere here my thoughts dry out..._
> * [00577f7](https://github.com/enso-org/enso/commit/00577f7c9664f6f21bcc1727dcf09e4c6f75507a) documents/tests the _chain of types_ and shows how we want to use `allTypes` method

- I am still relatively satisfied with 00577f7
   - e.g. there are no "static" methods - those are instance methods on `Type.getEigentype()`
- the only problem is the "duality of methods from Any"
- and I believe it is sufficiently solved by saying that:
   - each method found on `Any` has `self` argument defaulted to the receiver
   - but it can be changed by `self=whatever`
- at the end this sounds like a consistent set of rules